### PR TITLE
Preserving index when transforming

### LINKED ISSUE #277: If the input data has a different index, the reverse transformed data may be out of order
### Environment Details

Please indicate the following details about the environment in which you found the bug:

* RDT version: Any
* Python version: Any
* Operating System: Any

### Error Description

If the `DataFrame` input into the`HyperTransformer` has an index that isn't in incremental order, the reverse transformed data might come back in the wrong order. Some transformers reset the index and when the `BaseTransformer` assigns the new columns it is sometimes out of order. This happens in the following lines because the rows of `columns_data` will reorder to match the index of data.
https://github.com/sdv-dev/RDT/blob/10a996eebc8577692b49afc16a9be3417324d9fa/rdt/transformers/base.py#L126-L130

### Steps to reproduce

```python3
data = pd.DataFrame({
    'integer': [1, 2, 1, 3],
    'float': [0.1, 0.2, 0.1, 0.1],
    'categorical': ['a', 'a', 'b', 'a'],
    'bool': [False, False, True, False],
    'names': ['Jon', 'Arya', 'Jon', 'Jon'],
})
ht = HyperTransformer(data_type_transformers={'categorical': CategoricalTransformer})
ht.fit(data)
transformed = ht.transform(data)
reversed = ht.reverse_transform(data)
```

### PR DESCRIPTION
**Pull request by amontanez24**
resolves #277

### CODE REVIEW

File: rdt/transformers/base.py [hunk 1]
```diff
@@ -124,6 +126,9 @@ def _get_columns_data(data, columns):
 
     @staticmethod
     def _set_columns_data(data, columns_data, columns):
+        if isinstance(columns_data, pd.Series) or isinstance(columns_data, pd.DataFrame):
```
Comment (amontanez24) [inline rdt/transformers/base.py:129]: Using `assign` didn't fix the problem. I also realized that `columns_data` can be a Series, DataFrame or numpy array. So it seemed easiest to just fix the index if `columns_data` has one.

Comment (amontanez24) [reply rdt/transformers/base.py]: @csala I timed both using `to_numpy` and this method, and this seemed to be slightly faster for all `DataFrame` sizes

**Review by csala — APPROVED** [review]
This looks good!

**Review by katxiao — APPROVED** [review]
looks good 👍

**Pull request merged**